### ISSUE TITLE
Regression for ESM imported classes

### ISSUE THREAD

**Issue by shanewholloway** [rollup/rollup#4184]
### Rollup Version

^2.49.0

### Operating System (or Browser)

Linux (Ubuntu 21-04)

### Node Version (if applicable)

v12; v16

### Link To Reproduction

https://replit.com/@shanewholloway/rollup-regression-v2490-swh-class-use#readme.md

### Expected Behaviour

I expect `rollup -c` to produce an compiled output or emit an error with help to resolve any issues. This works as expected with rollup version `v2.48.0`.

### Actual Behaviour

`rollup -c` fails with version `v2.49` and after; no helpful error message to resolve issues.

- rollup `v2.49.0` to `v2.52.7` errors with `RangeError: Maximum call stack size exceeded`
- rollup `v2.52.8` to `v2.53.3` runs out of memory and core dumps on Linux.

**Comment by kzc** [comment]
Appears to be an infinite loop of calls to deoptimizeThisOnEventAtPath.

repro:
```
$ cat c.js
class Foo {
    static echo(message) {
        this.prototype.echo(message);
    }
    echo(message) {
        console.log(message);
    }
}

class Bar extends Foo {
}

global.baz = "PASS";
Bar.echo(baz);
```
expected:
```
$ cat c.js | node
PASS
```
actual:
```
$ rollup c.js --silent
...
FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
```

**Comment by lukastaegert** [comment]
Thanks a lot for the repro and refinement! I was able to write a small fix but it will need to wait until tomorrow to be refined and rolled out unfortunately.

**Comment by lukastaegert** [comment]
Fix at #4189

**Comment by shanewholloway** [comment]
v2.54.0 works great -- thanks!
